src/alarm_heap.cpp

```diff
@@ -0,0 +1,326 @@
+* the program along with SSL, set forth below. This program is distributed
+* in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+* without even the implied warranty of MERCHANTABILITY or FITNESS FOR
+* A PARTICULAR PURPOSE. See the GNU General Public License for more
+* details. You should have received a copy of the GNU General Public
+* License along with this program. If not, see
+* <http://www.gnu.org/licenses/>.
+*
+* The author can be reached by email at clearwater@metaswitch.com or by
+* post at Metaswitch Networks Ltd, 100 Church St, Enfield EN2 6BQ, UK
+*
+* Special Exception
+* Metaswitch Networks Ltd grants you permission to copy, modify,
+* propagate, and distribute a work formed by combining OpenSSL with The
+* Software, or a work derivative of such a combination, even if such
+* copying, modification, propagation, or distribution would otherwise
+* violate the terms of the GPL. You must comply with the GPL in all
+* respects for all of the code used other than OpenSSL.
+* "OpenSSL" means OpenSSL toolkit software distributed by the OpenSSL
+* Project and licensed under the OpenSSL Licenses, or a work based on such
+* software and licensed under the OpenSSL Licenses.
+* "OpenSSL Licenses" means the OpenSSL License and Original SSLeay License
+* under which the OpenSSL Project distributes the OpenSSL toolkit software,
+* as those licenses appear in the file LICENSE-OPENSSL.
+*/
+
+#include <stdexcept>
+#include <time.h>
+#include <cmath>
+#include <arpa/inet.h>
+
+#include "log.h"
+#include "alarm_heap.hpp"
+#include "itu_alarm_table.hpp"
+#include "alarm_active_table.hpp"
+
+AlarmInfo::~AlarmInfo()
+{
+  delete _alarm_in_heap; _alarm_in_heap = NULL;
+}
+
+AlarmSeverityChange AlarmInfo::new_alarm_severity(AlarmDef::Severity severity)
+{
+  // Order the severities so we can do a simple comparison to determine any
+  // severity change.
+  unsigned int ordered_severities[] = {0, 1, 2, 6, 5, 4, 3};
+  int old_severity = ordered_severities[_severity];
+  int new_severity = ordered_severities[severity];
+
+  if (old_severity == new_severity)
+  {
+    return AlarmSeverityChange::NO_CHANGE;
+  }
+  else if (old_severity > new_severity)
+  {
+    return AlarmSeverityChange::REDUCED_SEVERITY;
+  }
+  else
+  {
+    return AlarmSeverityChange::INCREASED_SEVERITY;
+  }
+}
+
+void AlarmInfo::set_severity(AlarmTableDef& alarm_table_def)
+{
+  // Update the severity in the map.
+  _severity = alarm_table_def.severity();
+
+  // Update the ActiveAlarmTable.
+  alarmActiveTable_trap_handler(alarm_table_def);
+}
+
+void AlarmInfo::update_alarm_in_heap(AlarmDef::Severity new_severity,
+                                     uint64_t time_to_delay)
+{
+  // Update the alarm in the heap's severity, and update its time to pop
+  // (which also rebalances the heap).
+  if (new_severity != _alarm_in_heap->severity())
+  {
+    _alarm_in_heap->set_severity(new_severity);
+    _alarm_in_heap->update_pop_time(time_to_delay);
```

Comment:
Are we ever going to want to update the pop time but with an unchanged severity? Worth a comment to say that's impossible?